quinn-proto/src/connection/mod.rs

```diff
@@ -2620,10 +2620,23 @@ where
                 Some(x) => x,
                 None => break,
             };
-            let len = cmp::min(
-                frame.data.len(),
-                max_size as usize - buf.len() - frame::Crypto::SIZE_BOUND,
-            );
+
+            // Calculate the maximum amount of crypto data we can store in the buffer.
+            // Since the offset is known, we can reserve the exact size required to encode it.
+            // For length we reserve 2bytes which allows to encode up to 2^14,
+            // which is more than what fits into normally sized QUIC frames.
+            let max_crypto_data_size = max_size
+                - buf.len()
+                - 1 // Frame Type
+                - VarInt::size(unsafe { VarInt::from_u64_unchecked(frame.offset) })
```

Comment:
Hmm, we should use `VarInt` more pervasively in frame structures.

Reply:
Agreed. But it would be better as a separate change